op/framework.py

```diff
@@ -374,14 +374,17 @@ def __init__(self, data_path, charm_dir, meta, model):
         self.charm_dir = charm_dir
         self.meta = meta
         self.model = model
-        self._event_count = 0
         self._observers = []      # [(observer_path, method_name, parent_path, event_key)]
         self._observer = weakref.WeakValueDictionary()       # {observer_path: observer}
         self._type_registry = {}  # {(parent_path, kind): cls}
         self._type_known = set()  # {cls}
 
         self._storage = SQLiteStorage(data_path)
 
+        # Note: we can't persist the event count using StoredData because it relies on events.
+        self._event_count_handle = self.handle.nest('_event_count', None)
+        self._event_count = self._storage.load_snapshot(self._event_count_handle.path) or 0
```

Comment:
2^31 is 1 event per second for 68 years. So if we get on the order of 10/100ms it starts to actually be a factor. 2^63 then we really never hit it. (1 event per microsecond takes ~300,000 years to overflow.) 

It would be interesting to have a forced mod(2,000,000,000) as then you don't overflow an int32 and then you don't have to worry about conversions/whether one bit of code stores it as int64 and another as int32.

Reply:
If the snapshot logic is unable to save larger Python ints properly, that sounds like an issue by itself which will blow up elsewhere. So I think we can trust that to work, and might even test whether it works, but that's likely unnecessary before we're doing more than just passing it through to pickle to handle it.